fix!: OpenID4VCI Draft 13 (#175)
italia/eidas-it-wallet-docs: docs/en/pid-eaa-issuance.rst

High-Level PID flow
-------------------

The :numref:`fig_High-Level-Flow-ITWallet-PID-Issuance` shows a general architecture and highlights the main operations involved in the issuance of a PID.

.. _fig_High-Level-Flow-ITWallet-PID-Issuance:
.. figure:: ../../images/High-Level-Flow-ITWallet-PID-Issuance.svg
    :figwidth: 100%
    :align: center

    PID Issuance - General architecture and high level flow.

Below the description of the steps represented in the previous picture:

    0. **Wallet Instance Setup**: the first time the Wallet Instance is started a preliminary setup phase is carried out. It consists of the release of the Wallet Instance Attestation issued by Wallet Attestation Service asserting the genuineness and the compliance of the Wallet Instance with the shared trust framework. The Wallet Instance Attestation binds the public key provided by the Wallet Instance, related to one of the private keys generated by the Wallet Instance.

>     1. **PID/(Q)EAA Provider Issuers Discovery**: the Wallet Instance discovers the trusted Digital Credential Issuers using the Federation API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), inspecting the Credential Issuer metadata and Trust Marks for filtering the PID Provider.

Reviewer: I think we have to keep only Provider or Issuer in the "PID/(Q)EAA Provider Issuers Discovery". Otherwise, it seems redundant to me. WDYT @fmarino-ipzs @peppelinux

Author: yes, mistake!

Reviewer: no worries.

>     1. **PID/(Q)EAA Provider Issuers Discovery**: the Wallet Instance discovers the trusted Digital Credential Issuers using the Federation API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), inspecting the Credential Issuer metadata and Trust Marks for filtering the PID Provider.

Author: ```suggestion
    1. **PID/(Q)EAA Provider Discovery**: the Wallet Instance discovers the trusted Digital Credential Issuers using the Federation API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), inspecting the Credential Issuer metadata and Trust Marks for filtering the PID Provider.
```

PID/(Q)EAA Issuance
+++++++++++++++++++

This section describes the PID and (Q)EAAs issuance flow with an high level of security.
The relevant entities and interfaces involved in the issuance flow are:

>     - *Wallet Provider*: The entity responsible for releasing an EUDI Wallet Solution. The Wallet Provider issues the Wallet Instance Attestations to its Wallet Instances through an Attestation Service. The Wallet Attestation certifies the genuinity and authenticity of the Wallet Instance and its compliance the security and privacy requirements.

Reviewer: ```suggestion
    - *Wallet Provider*: The entity responsible for releasing an EUDI Wallet Solution. The Wallet Provider issues the Wallet Instance Attestations to its Wallet Instances through an Attestation Service. The Wallet Attestation certifies the genuinity and authenticity of the Wallet Instance and its compliance with the security and privacy requirements.
```

>     - *Wallet Provider*: The entity responsible for releasing an EUDI Wallet Solution. The Wallet Provider issues the Wallet Instance Attestations to its Wallet Instances through an Attestation Service. The Wallet Attestation certifies the genuinity and authenticity of the Wallet Instance and its compliance with a trust framework in compliance to the security and privacy requirements.

Author: ```suggestion
    - *Wallet Provider*: The entity responsible for releasing an EUDI Wallet Solution. The Wallet Provider issues the Wallet Instance Attestations to its Wallet Instances through an Attestation Service. The Wallet Attestation certifies the genuinity and authenticity of the Wallet Instance and its compliance the security and privacy requirements.
```

    - *Wallet Instance*: Instance of a Wallet Solution, installed on the User device. The Wallet Instance provides graphical interfaces for User interaction with Relying Parties, PID, (Q)EAA Providers and the Wallet Provider.
    - *PID Provider*: The entity that issues the eIDAS Person Identification Data (PID). It is composed of:

        - OpenID4VCI Component: based on the "OpenID for Verifiable Credential Issuance" specification  ` [OIDC4VCI. Draft 13] <https://openid.bitbucket.io/connect/openid-4-verifiable-credential-issuance-1_0.html>`_ to release the PID.
        - National eID Relying Party: The component to authenticate the User with the national Digital Identity Providers, based on OpenID Connect Core 1.0 or SAML2.
        - National Identity Provider: It represents preexisting identity systems based on SAML2 or OpenID Connect Core 1.0, already in production in each Member State (eg: the Italian SPID and CIE id schemes notified eIDAS with *LoA* **High**, see `SPID/CIE OpenID Connect Specifications <https://italia.github.io/spid-cie-oidc-docs/en/>`_).

    - *(Q)EAA Provider*: It represents the Issuer of (Q)EAAs. It is composed of:

>       - OpenID4VCI Component: based on the "OpenID for Verifiable Credential Issuance" specification ` to release (Q)EAAs.

Reviewer: ```suggestion
      - OpenID4VCI Component: based on the "OpenID for Verifiable Credential Issuance" specification to release (Q)EAAs.
```

High-Level PID flow
-------------------

The :numref:`fig_High-Level-Flow-ITWallet-PID-Issuance` shows a general architecture and highlights the main operations involved in the issuance of a PID.

.. _fig_High-Level-Flow-ITWallet-PID-Issuance:
.. figure:: ../../images/High-Level-Flow-ITWallet-PID-Issuance.svg
    :figwidth: 100%
    :align: center

    PID Issuance - General architecture and high level flow.

Below the description of the steps represented in the previous picture:

    0. **Wallet Instance Setup**: the first time the Wallet Instance is started a preliminary setup phase is carried out. It consists of the release of the Wallet Instance Attestation issued by Wallet Attestation Service asserting the genuineness and the compliance of the Wallet Instance with the shared trust framework. The Wallet Instance Attestation binds the public key provided by the Wallet Instance, related to one of the private keys generated by the Wallet Instance.

>     1. **Digital Credential Issuers Discovery**: the Wallet Instance discovers the trusted Digital Credential Issuers using the Federation API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), inspecting the Credential Issuer metadata and Trust Marks for filtering the PID Provider.

Reviewer: As we are in the section specific to the PID and (Q)EAA I would keep the PID/(Q)EAA Provider term rather than digital credential issuer discovery. WDYT?

>     1. **Digital Credential Issuers Discovery**: the Wallet Instance discovers the trusted Digital Credential Issuers using the Federation API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), inspecting the Credential Issuer metadata and Trust Marks for filtering the PID Provider.

Author: ```suggestion
    1. **PID/(Q)EAA Provider Issuers Discovery**: the Wallet Instance discovers the trusted Digital Credential Issuers using the Federation API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), inspecting the Credential Issuer metadata and Trust Marks for filtering the PID Provider.
```

Below the description of the most relevant operations involved in the (Q)EAA issuance:

>     1. **Discovery of the trusted (Q)EAA Issuer**: the Wallet Instance obtains the list of the trusted (Q)EAA Issuer using the Federaiton API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), then inspects the metadata and Trust Mark looking for the Digital Credential capabilities of each (Q)EAA Provider.

Reviewer: ```suggestion
    1. **Discovery of the trusted (Q)EAA Provider**: the Wallet Instance obtains the list of the trusted (Q)EAA Provider using the Federation API (e.g.: using the Subordinate Listing Endpoint of the Trust Anchor and its Intermediates), then inspects the metadata and Trust Mark looking for the Digital Credential capabilities of each (Q)EAA Provider.
```

  * specifies the types of the requested credentials using the ``authorization_details`` [RAR :rfc:`9396`] parameter.

The PID/(Q)EAA Provider performs the following checks upon the receipt of the PAR request:

    1. It MUST validate the signature of the Request Object using the algorithm specified in the ``alg`` header parameter (:rfc:`9126`, :rfc:`9101`) and the public key retrieved from the Wallet Instance Attestation (``cnf.jwk``) referenced in the Request Object, using the ``kid`` JWS header parameter.
    2. It MUST check that the used algorithm for signing the request in the ``alg`` header is one of the listed within the Section `Cryptographic Algorithms <algorithms.html>`_.
    3. It MUST check that the ``client_id`` in the request body of the PAR request matches the ``client_id`` claim included in the Request Object.
    4. It MUST check that the ``iss`` claim in the Request Object matches the ``client_id`` claim in the Request Object (:rfc:`9126`, :rfc:`9101`).

>     5. It MUST check that the ``aud`` claim in the Request Object is equal to the PID/(Q)EAA Provider authorization endpoint uri (:rfc:`9126`, :rfc:`9101`).

Reviewer 2: > The JWT MUST contain an "aud" (audience) claim containing a value that identifies the authorization server as an intended audience. The [[RFC8414](https://www.ietf.org/archive/id/draft-ietf-oauth-attestation-based-client-auth-01.html#RFC8414)] issuer identifier URL of the authorization server MUST be used as a value for an "aud" element to identify the authorization server as the intended audience of the JWT. [Attestation-based Draft](https://www.ietf.org/archive/id/draft-ietf-oauth-attestation-based-client-auth-01.html)
As it is mentioned in the draft, it should be the URL of the Authorization Server and not the Authorization endpoint

    POST /as/par HTTP/1.1
    Host: pid-provider.example.org
    Content-Type: application/x-www-form-urlencoded

    response_type=code
    &client_id=$thumprint-of-the-jwk-in-the-cnf-wallet-attestation$
    &code_challenge=E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM
    &code_challenge_method=S256
    &request=eyJhbGciOiJSUzI1NiIsImtpZCI6ImsyYmRjIn0.ew0KIC Jpc3MiOiAiczZCaGRSa3F0MyIsDQogImF1ZCI6ICJodHRwczovL3NlcnZlci5leGFtcGxlLmNvbSIsDQo gInJlc3BvbnNlX3R5cGUiOiAiY29kZSBpZF90b2tlbiIsDQogImNsaWVudF9pZCI6ICJzNkJoZFJrcXQz IiwNCiAicmVkaXJlY3RfdXJpIjogImh0dHBzOi8vY2xpZW50LmV4YW1...
    &client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-client-attestation
    &client_assertion=$WIA~WIA-PoP

Below an non-normative example of the Wallet Instance Attestation Proof of Possession without encoding and signature applied:

.. code-block::

  {
    "alg": "ES256",
    "kid": "vbeXJksM45xphtANnCiG6mCyuU4jfGNzopGuKvogg9c",
    "typ": "jwt-client-attestation-pop",
  }
  .
  {
    "iss": "vbeXJksM45xphtANnCiG6mCyuU4jfGNzopGuKvogg9c",

>     "aud": "https://pid-provider.example.org/authorization-endpoint",

Reviewer 3: ```suggestion
    "aud": "https://pid-provider.example.org/par-endpoint",
```

  }
  .
  {
    "iss": "$thumprint-of-the-jwk-in-the-cnf-wallet-attestation$",
    "client_id": "$thumprint-of-the-jwk-in-the-cnf-wallet-attestation$",
    "exp": 1672422065,
    "iat": 1672418465,
    "aud": "https://pid-provider.example.org/authorization-endpoint",
    "jti": "ac80df576e7109686717bf50b869e882",
    "response_type": "code",
    "state": "fyZiOL9Lf2CeKuNT2JzxiLRDink0uPcd",
    "code_challenge": "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM",
    "code_challenge_method": "S256",
    "authorization_details": [
      {

>           "type": "openid_credential",
>           "format": "vc+sd-jwt",
>           "credential_definition": {
>               "type": ["PersonIdentificationData"]

Author: ```suggestion
          "type": "openid_credential",
          "credential": "PersonIdentificationData"
```

Author: related to https://github.com/openid/OpenID4VCI/pull/137

.. note::

    The Wallet Instance redirect URI is a universal or app link registered with the local operating system, so this latter will resolve it and pass the response to the Wallet Instance.

.. code-block:: http

    HTTP/1.1 302 Found
    Location: https://start.wallet.example.org?code=SplxlOBeZQQYbYS6WxSbIA&state=fyZiOL9Lf2CeKuNT2JzxiLRDink0uPcd&iss=https%3A%2F%2Fpid-provider.example.org

**Steps 12-13 (DPoP Proof for Token Endpoint)**: The Wallet Instance MUST create a new key pair for the DPoP and a fresh DPoP Proof JWT following the instruction provided in Section 4 of (:rfc:`9449`) for the token request to the PID/(Q)EAA Provider. The DPoP Proof JWT is signed using the private key for DPoP created by Wallet Instance for this scope. DPoP binds the Access Token to a certain Wallet Instance (:rfc:`9449`) and mitigates the misuse of leaked or stolen Access Tokens at the Credential Endpoint.

> **Step 14 (Token Request):** The Wallet Instance sends a token request to the PID/(Q)EAA Provider Token Endpoint using the authorization ``code``, ``code_verifier`` and *DPoP Proof JWT*. The PID/(Q)EAA Provider performs the following checks:

Author: ```suggestion
**Step 14 (Token Request):** The Wallet Instance sends a token request to the PID/(Q)EAA Provider Token Endpoint with a *DPoP Proof JWT*  and the parameters: ``code``, ``code_verifier``, and OAuth 2.0 Attestation based Client Authentication  (``client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-client-attestation`` and ``client_assertion=WIA~WIA-PoP``). 
The ``client_assertion`` is signed using the private key that is created during the setup phase to obtain the Wallet Instance Attestation. The related public key that is attested by the Wallet Provider is provided within the Wallet Instance Attestation (``cnf`` claim). The PID/(Q)EAA Provider performs the following checks on the Token Request:
```

**Steps 12-13 (DPoP Proof for Token Endpoint)**: The Wallet Instance MUST create a new key pair for the DPoP and a fresh DPoP Proof JWT following the instruction provided in Section 4 of (:rfc:`9449`) for the token request to the PID/(Q)EAA Provider. The DPoP Proof JWT is signed using the private key for DPoP created by Wallet Instance for this scope. DPoP binds the Access Token to a certain sender (Wallet Instance) (:rfc:`9449`) and mitigates the misuse of leaked or stolen Access Tokens at the Credential Endpoint of PID/(Q)EAA Issuer.

**Step 14 (Token Request):** The Wallet Instance sends a token request to the PID/(Q)EAA Provider Token Endpoint using the authorization ``code``, ``code_verifier``, *DPoP Proof JWT* and OAuth 2.0 Attestation based Client Authentication parameters (``client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-client-attestation`` and ``client_assertion=WIA~WIA-PoP``). 

> The ``client_assertion`` is signed using the private key that is created during the setup phase to obtain the Wallet Instance Attestation. The related public key that is attested by the Wallet Provider is provided within the Wallet Instance Attestation (``cnf`` claim). The PID/(Q)EAA Provider performs the following checks on the Token Request:

Author: well, considering that at this endpoint the client uses the code, obtained after a client and an user authentication and also a code challenge (that's a secret!), I'd keep the WIA~POP at protection of the authorization endpoint without requiring it again here.

@asharif1990 ^

Reviewer: So, I would recommend keeping it here in any case as an additional security layer. It is right before releasing the Access Token for the credential endpoint so keeping the Wallet Authentication would not harm I would say. The Wallet Authentication at the PAR endpoint is an early-stage mitigation, while the authentication at the token endpoint is a good practice to be kept and to be compliant with both Core specs and HAIP. WDYT?

Author: I think that I'll keep it as it is with WIA~POP at the token endpoint to not increase the noise, with the awareness that we have discussed this and that we'll continue exploring this way with other authors to get an agreement

Author: text restored and comment resolved, @asharif1990 ^

        WF0IjoxNTYyMjYyNjE2fQ.2-GxA6T8lP4vfrg8v-FdWP0A0zdrj8igiMLvqRMUvwnQg
        4PtFLbdLXiOSsX0x7NVY-FNyJK70nfbV37xRZT3Lg

    grant_type=authorization_code
    &code=SplxlOBeZQQYbYS6WxSbIA
    &code_verifier=dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk
    &redirect_uri=https://start.wallet.example.org/cb

> 

Reviewer: ```suggestion
    &client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-client-attestation
    &client_assertion=$WIA~WIA-PoP
```

        WF0IjoxNTYyMjYyNjE2fQ.2-GxA6T8lP4vfrg8v-FdWP0A0zdrj8igiMLvqRMUvwnQg
        4PtFLbdLXiOSsX0x7NVY-FNyJK70nfbV37xRZT3Lg

    grant_type=authorization_code
    &code=SplxlOBeZQQYbYS6WxSbIA
    &code_verifier=dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk
    &redirect_uri=https://start.wallet.example.org/cb

**Step 15 (Token Response):** The PID/(Q)EAA Provider validates the request, if successful an *Access Token* (bound to the DPoP key) and a fresh ``c_nonce`` is provided to the Wallet Instance.

.. code-block:: http

    HTTP/1.1 200 OK
    Content-Type: application/json
    Cache-Control: no-store

    {
        "access_token": "Kz~8mXK1EalYznwH-LC-1fBAo.4Ljp~zsPE_NeO.gxU ...",
        "token_type": "DPoP",
        "expires_in": 2677,
        "c_nonce": "tZign[...]snFbp",
        "c_nonce_expires_in": 86400,
        "authorization_details": [
          {
              "type": "openid_credential",

>               "format": "vc+sd-jwt",
>               "credential_definition": {
>                   "type": ["PersonIdentificationData"]

Author: ```suggestion
              "credential": "PersonIdentificationData"
```

Author: according to https://github.com/openid/OpenID4VCI/pull/137

  }

> Where the decoded content of the ``jwt`` parameter is represented below,

Author: ```suggestion
Where a non-normative example of the decoded content of the ``jwt`` parameter is represented below,
```

> without endpding and signature, for example purpose. The JWS header:

Author: ```suggestion
without encoding and signature. The JWS header:
```

  }

And the JWS payload:

.. code-block:: JSON

    {
        "iss": "0b434530-e151-4c40-98b7-74c75a5ef760",

>         "aud": "https://pid-provider.example.org",

Reviewer 3: ```suggestion
        "aud": "https://pid-provider.example.org/credential",
```

  }

> And the The JWS payload:

Author: ```suggestion
And the JWS payload:
```

.. code-block:: JSON

    {
        "iss": "0b434530-e151-4c40-98b7-74c75a5ef760",
        "aud": "https://pid-provider.example.org",
        "iat": 1504699136,
        "nonce": "tZign...snFbp"
    }

> **Steps 19-21 (Credential Response):** The PID/(Q)EAA Provider MUST validate the *DPoP JWT Proof* based on the steps defined in Section 4.3 of (:rfc:`9449`) and whether the *Access Token* is valid and suitable for the requested PID/(Q)EAA. It also MUST validate the proof of possession for the key material the new credential SHALL be bound to following the steps in Section 7.2.2 of `OPENID4VCI`_. If all checks succeed, the PID/(Q)EAA Provider creates a new Credential bound to the key material and provide it to the Wallet Instance. The Wallet Instance MUST perform the following checks before proceeding with the secure storage of the PID/(Q)EAA:

Author: ```suggestion
**Steps 19-21 (Credential Response):** The PID/(Q)EAA Provider MUST validate the *DPoP JWT Proof* based on the steps defined in Section 4.3 of (:rfc:`9449`) and whether the *Access Token* is valid and suitable for the requested PID/(Q)EAA. It also MUST validate the proof of possession for the key material the new credential SHALL be bound to, according to `OPENID4VCI`_ Section 7.2.2. If all checks succeed, the PID/(Q)EAA Provider creates a new Credential bound to the key material and provide it to the Wallet Instance. The Wallet Instance MUST perform the following checks before proceeding with the secure storage of the PID/(Q)EAA:
```

    1. It MUST check that the PID Credential Response contains all the mandatory parameters and values are validated according to :ref:`Table of the credential response parameters <table_credential_response_claim>`.
    2. It MUST check the PID integrity by verifying the signature using the algorithm specified in the ``alg`` header parameter of SD-JWT (:ref:`PID/(Q)EAA Data Model <pid_eaa_data_model.rst>`) and the public key that is identified using using the ``kid`` header of the SD-JWT.
    3. It MUST check that the received PID (in credential claim) matches the schema defined in :ref:`PID/(Q)EAA Data Model <pid_eaa_data_model.rst>`.
    4. It MUST process and verify the PID in SD-JWT VC format (according to `SD.JWT#Verification <https://drafts.oauth.net/oauth-selective-disclosure-jwt/draft-ietf-oauth-selective-disclosure-jwt.html#name-verification-and-processing>`_ Section 6.) or MDOC CBOR format. 
    5. It MUST verify the Trust Chain in the header of SD-JWT VC to verify that the PID Provider is trusted.

> If the checks defined above are successful the Wallet Instance proceeds with the secure storage of the Digital Credential.

Reviewer: as I mentioned in my previous comment maybe in the intro section we can use Digital Credential, but then as we are speaking about the release of PID and (Q)EAA we can keep them.

> If the checks defined above are successful the Wallet Instance proceeds with the secure storage of the Digital Credential.

Author: ```suggestion
If the checks defined above are successful the Wallet Instance proceeds with the secure storage of the PID/(Q)EAA.
```

      - Unique identifier of the JWT that, together with the value contained in the ``iss`` claim,  prevents the reuse of the JWT (replay attack). Since the `jti` value alone is not collision resistant, it MUST be identified uniquely together with its issuer.
      - [:rfc:`7519`].

> The JOSE header of the Wallet Instance Attestation Proof of Possession MUST contain:

Author: ```suggestion
The JOSE header of the Wallet Instance Attestation proof of possession MUST contain:
```

      - :rfc:`7516#section-4.1.1`.
    * - **kid**
      -  Unique identifier of the ``jwk`` inside the ``cnf`` claim of Wallet Instance Attestation as base64url-encoded JWK Thumbprint value.
      - :rfc:`7638#section_3`.
    * - **typ**
      -  It MUST be set to ``jwt-client-attestation-pop``
      -  Currently under discussion in [`oauth-attestation-draft <https://vcstuff.github.io/draft-ietf-oauth-attestation-based-client-auth/draft-ietf-oauth-attestation-based-client-auth.html>`_].

> The body of the Wallet Instance Attestation Proof of Possession JWT MUST contain:

Author: ```suggestion
The body of the Wallet Instance Attestation proof of possession JWT MUST contain:
```

      - `OAuth 2.0 Authorization Server Issuer Identifier in Authorization Response <https://www.ietf.org/archive/id/draft-meyerzuselhausen-oauth-iss-auth-resp-02.html>`_, `[RFC7519, Section 4.1.1] <https://www.iana.org/go/rfc7519>`_.

Token endpoint
--------------

> The token endpoint is used by the Wallet Instance to obtain an Access Token by presenting its authorization grant, as

Author: ```suggestion
The token endpoint is used by the Wallet Instance to obtain an Access Token by presenting an authorization grant, as
```

defined in :rfc:`6749`. The Token Endpoint is a protected endpoint with a client authentication based on the model defined in OAuth 2.0 Attestation-based Client Authentication [`oauth-attestation-draft <https://vcstuff.github.io/draft-ietf-oauth-attestation-based-client-auth/draft-ietf-oauth-attestation-based-client-auth.html>`_].

Token Request
^^^^^^^^^^^^^^^

The request to the PID/(Q)EAA Token endpoint MUST be an HTTP request with method POST, with the body message encoded in ``application/x-www-form-urlencoded`` format. The Wallet Instance sends the Token endpoint request with ``client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-client-attestation`` and ``client_assertion=WIA~WIA-PoP``.

> The Token endpoint MUST accept and validate the DPoP proof sent in the DPoP HTTP header. The Token endpoint MUST validate the DPoP proof according to Section 4.3 of the DPoP specifications (:rfc:`9449`). Thus, this mitigates the misuse of leaked or stolen Access Tokens at the credential endpoint. If the DPoP proof is invalid, the Token endpoint returns an error response, according to Section 5.2 of [:rfc:`6749`] with ``invalid_dpop_proof`` as the value of the error parameter.

Author: ```suggestion
The Token endpoint MUST accept and validate the DPoP proof sent in the DPoP HTTP header. The Token endpoint MUST validate the DPoP proof according to Section 4.3 of the DPoP specifications (:rfc:`9449`). This mitigates the misuse of leaked or stolen Access Tokens at the credential endpoint. If the DPoP proof is invalid, the Token endpoint returns an error response, according to Section 5.2 of [:rfc:`6749`] with ``invalid_dpop_proof`` as the value of the error parameter.
```

    * - **code_verifier**
      - Verification code of the **code_challenge**.
      - `Proof Key for Code Exchange by OAuth Public Clients <https://datatracker.ietf.org/doc/html/rfc7636>`_.

A **DPoP Proof JWT** is included in the HTTP request using the ``DPoP`` header parameter containing a DPoP JWS.

The JOSE header of a **DPoP JWT** MUST contain at least the following parameters:

.. list-table::
    :widths: 20 60 20
    :header-rows: 1

    * - **JOSE header**
      - **Description**
      - **Reference**
    * - **typ**
      - It MUST be equal to ``dpop+jwt``.
      - [:rfc:`7515`] and [:rfc:`8725`. Section 3.11].
    * - **alg**
      - A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST be one of the supported algorithms in Section :ref:`Cryptographic Algorithms <supported_algs>` and MUST NOT be set to ``none`` or with a symmetric algorithm (MAC) identifier.
      - [:rfc:`7515`].
    * - **jwk**

>       - It represents the public key chosen by the Wallet Instance, in JSON Web Key (JWK) [:rfc:`7517`] format that the Access Token SHALL be bound to, as defined in [:rfc:`7515`] Section 4.1.3. It MUST NOT contain a private key.

Reviewer 3: ```suggestion
      - It represents the public key chosen by the Wallet Instance, in JSON Web Key (JWK) [:rfc:`7517`] format that the Access Token MUST be bound to, as defined in [:rfc:`7515`] Section 4.1.3. It MUST NOT contain a private key.
```

    * - **code_verifier**
      - Verification code of the **code_challenge**.
      - `Proof Key for Code Exchange by OAuth Public Clients <https://datatracker.ietf.org/doc/html/rfc7636>`_.

A **DPoP Proof JWT** is included in the HTTP request using the ``DPoP`` header parameter containing a DPoP JWS.

The JOSE header of a **DPoP JWT** MUST contain at least the following parameters:

.. list-table::
    :widths: 20 60 20
    :header-rows: 1

    * - **JOSE header**
      - **Description**
      - **Reference**
    * - **typ**
      - It MUST be equal to ``dpop+jwt``.
      - [:rfc:`7515`] and [:rfc:`8725`. Section 3.11].
    * - **alg**
      - A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST be one of the supported algorithms in Section :ref:`Cryptographic Algorithms <supported_algs>` and MUST NOT be set to ``none`` or with a symmetric algorithm (MAC) identifier.
      - [:rfc:`7515`].
    * - **jwk**

>       - It represents the public key chosen by the Wallet Instance, in JSON Web Key (JWK) [:rfc:`7517`] format that the Access Token SHALL be bound to, as defined in Section 4.1.3 of [:rfc:`7515`]. It MUST NOT contain a private key.

Author: ```suggestion
      - It represents the public key chosen by the Wallet Instance, in JSON Web Key (JWK) [:rfc:`7517`] format that the Access Token SHALL be bound to, as defined in [:rfc:`7515`] Section 4.1.3. It MUST NOT contain a private key.
```

      - Expiry time of the *Access Token* in seconds.
      - :rfc:`6749`.
    * - **c_nonce**

>       - JSON string containing a ``nonce`` value to be used to create a *proof of possession* of key material when requesting a Credential.

Reviewer 3: If I'm not mistaken, Credential is not a defined term. I see only Digital Credential. Probably we should say in the defined term Section that Credential refers to Digital Credential. 
@peppelinux WDYT?

Author: The text that defines the term DigitaleCredential also says 

> It is also known as Credential.

      - Expiry time of the *Access Token* in seconds.
      - :rfc:`6749`.
    * - **c_nonce**

>       - JSON string containing a ``nonce`` value to be used to create a *proof of possession* of key material when requesting a credential.

Author: ```suggestion
      - JSON string containing a ``nonce`` value to be used to create a *proof of possession* of key material when requesting a Credential.
```
